// @shopify/cli-kit/dist/api/oxygen.js
import { buildHeaders, debugLogRequest } from './common.js';
import { graphqlClient } from '../http/graphql.js';
import { shopifyFetch } from '../http.js';
export async function request(oxygenAddress, query, token, variables) {
    const headers = await buildHeaders(token);
    debugLogRequest('Oxygen', query, variables, headers);
    const client = await graphqlClient({
        headers,
        url: getOxygenAddress(oxygenAddress),
    });
    const response = await client.request(query, variables);
    return response;
}
export async function uploadDeploymentFile(oxygenAddress, token, data) {
    const headers = await buildHeaders(token);
    delete headers['Content-Type'];
    const response = await shopifyFetch(getOxygenAddress(oxygenAddress), {
        method: 'POST',
        body: data,
        headers: {
            ...headers,
            ...data.getHeaders(),
        },
    });
    return response;
}
const getOxygenAddress = (oxygenHost) => {
    return `https://${oxygenHost}/api/graphql/deploy/v1`;
};
