// @shopify/app/dist/cli/commands/app/build.js
import { appFlags } from '../../flags.js';
import { load as loadApp } from '../../models/app/loader.js';
import build from '../../services/build.js';
import Command from '../../utilities/app-command.js';
import { Flags } from '@oclif/core';
import { path, cli, metadata } from '@shopify/cli-kit';
export default class Build extends Command {
    async run() {
        const { flags } = await this.parse(Build);
        await metadata.addPublic(() => ({
            cmd_app_dependency_installation_skipped: flags['skip-dependencies-installation'],
        }));
        const directory = flags.path ? path.resolve(flags.path) : process.cwd();
        const app = await loadApp(directory);
        await build({ app, skipDependenciesInstallation: flags['skip-dependencies-installation'], apiKey: flags['api-key'] });
    }
}
Build.description = 'Build the app';
Build.flags = {
    ...cli.globalFlags,
    ...appFlags,
    'skip-dependencies-installation': Flags.boolean({
        hidden: false,
        description: 'Skips the installation of dependencies.',
        env: 'SHOPIFY_FLAG_SKIP_DEPENDENCIES_INSTALLATION',
        default: false,
    }),
    'api-key': Flags.string({
        hidden: false,
        description: "Application's API key that will be exposed at build time.",
        env: 'SHOPIFY_FLAG_API_KEY',
    }),
};
